src/Symfony/Component/Messenger/RoutableMessageBus.php

```diff
@@ -40,29 +40,32 @@ public function dispatch($envelope, array $stamps = []): Envelope
             throw new InvalidArgumentException('Messages passed to RoutableMessageBus::dispatch() must be inside an Envelope');
         }
 
-        /** @var BusNameStamp|null $busNameStamp */
-        $busNameStamp = $envelope->last(BusNameStamp::class);
+        /** @var ReceivedStamp|null $receivedStamp */
+        $receivedStamp = $envelope->last(ReceivedStamp::class);
 
-        if (null === $busNameStamp) {
-            if (null === $this->fallbackBus) {
-                throw new InvalidArgumentException(sprintf('Envelope is missing a BusNameStamp and no fallback message bus is configured on RoutableMessageBus.'));
-            }
+        if (null === $receivedStamp) {
+            // The RoutableMessageBus is only used in Worker context where the ReceivedStamp is added. So this should not happen.
+            throw new InvalidArgumentException('Envelope is missing a ReceivedStamp.');
+        }
 
-            return $this->fallbackBus->dispatch($envelope, $stamps);
+        /** @var SentToFailureTransportStamp|null $sentToFailureStamp */
+        $sentToFailureStamp = $envelope->last(SentToFailureTransportStamp::class);
```

Comment:
This part replaces the `FailedMessageProcessingMiddleware` and makes sure the correct bus is used for messages consumed from the failure transport (directly or via messenger:failed:retry)